docs/design/rosbag2_snapshot_recorder/rosbag2_snapshot_recorder.md

```diff
@@ -0,0 +1,738 @@
+
+* Save data to rosbag on trigger
+  * Enable triggering buffer save to disk with service call
+
+* Pause and resume buffering process
+  * Pausing refers to stopping buffering new messages until resumed or write is triggered.
+  * Resuming continues the process of buffering after pause. Resuming would clear all buffers.
+
+* Node and buffer configuration
+
+  Enable user to specify `cli` arguments which could be shared with `ros2 bag record` command. User should be able to configure the following arguments:
+
+    | parameter name | description   | behavior  |
+    | -------------- | ------------- | --------- |
+    | buffer size    | maximum memory per topic | <ul><li> defined by new `--max-buffer-size` argument, default value is 0 [bytes]  </li><li>if not set/set to non-positive value only duration parameter will be relevant (*note: either buffer size or duration must be specified*) </li></ul> |
+    | duration       | maximum time difference between newest and oldest message | <ul><li> defined by new `--max-buffer-duration` argument, default value is 10.0 [s] </li><li> if not set/set to non-positive value only buffer size parameter will be relevant (*note: either buffer size or duration must be specified*) </li></ul> |
+    | subscribed topics | list of topics to buffer | <ul><li> makes use of `topics`, `--regex`, `--exclude`, `--all` arguments from `ros2 bag record` </li><li> if a topic is not available it won't be recorded </li></ul> |
+    | output rosbag filename | name of snapshot file saved to disk | <ul><li> defined by existing `--output` argument </li><li> rosbag name will be suffixed with timestamp data to prevent name collisions after triggering multiple snapshots </li><li> if not specified, default, timestamped format name will be used </li></ul> |
+
+
+## User interface
+
+This section shows how user can interact with the following feature.
+
+```bash
+# Running Recorder node in buffer mode for all topics with default configuration
+ros2 bag record --snapshot --all
+
+# Running Recorder node in buffer mode with all snapshot specific arguments
+ros2 bag record --snapshot --max-buffer-size 5000000 --max-buffer-duration 30.0 --output snapshot /example_topic1 /example_topic2
+
+# Triggering a snapshot with a service call
+ros2 service call /rosbag2_recorder/snapshot rosbag2_interfaces/srv/Snapshot "{ }"
+
+# Reconfigure snapshot max memory configuration
+ros2 service call /rosbag2_recorder/snapshot/memory rosbag2_interfaces/srv/SetMemory "{memory: 1000000}"
+
+# Reconfigure snapshot max duration configuration
+ros2 service call /rosbag2_recorder/snapshot/duration rosbag2_interfaces/srv/SetDuration "{duration: 50.0}"
+
+# Clear snapshot buffer
+ros2 service call /rosbag2_recorder/snapshot/clear_buffer rosbag2_interfaces/srv/ClearBuffer "{ }"
```

Comment:
I'm a little confused by this feature (and it doesn't seem to be mentioned in the introduction) - a snapshot write should clear the buffers, but why would we want to do it without a write?